include/pybind11/iostream.h

```diff
@@ -0,0 +1,188 @@
+#pragma once
+
+#include "pybind11.h"
+
+#include <streambuf>
+#include <ostream>
+#include <string>
+#include <memory>
+#include <iostream>
+
+NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+NAMESPACE_BEGIN(detail)
+
+// Buffer that writes to Python instead of C++
+class pythonbuf : public std::streambuf {
+private:
+    using traits_type = std::streambuf::traits_type;
+    char          d_buffer[1024];
+
+    object pywrite;
+    object pyflush;
+
+    int overflow(int c) {
+        if (!traits_type::eq_int_type(c, traits_type::eof())) {
+            *this->pptr() = traits_type::to_char_type(c);
+            this->pbump(1);
+        }
+        return this->sync() ? traits_type::not_eof(c) : traits_type::eof();
+    }
+
+    int sync() {
+        if (this->pbase() != this->pptr()) {
+            // This subtraction cannot be negative, so dropping the sign
+            str line(this->pbase(), static_cast<size_t>(this->pptr() - this->pbase()));
+
+            pywrite(line);
+            pyflush();
+
+            this->setp(this->pbase(), this->epptr());
+        }
+        return 0;
+    }
+public:
+    pythonbuf(object pyostream)
+        : pywrite(pyostream.attr("write")),
+          pyflush(pyostream.attr("flush")) {
+        this->setp(this->d_buffer, this->d_buffer + sizeof(this->d_buffer) - 1);
+    }
+};
+
+// Stream that writes to Python instead of C++ file descriptors
+class opythonstream : public std::ostream {
+public:
+    opythonstream(object pyostream) : std::ostream(new pythonbuf(pyostream)) { }
+    virtual ~opythonstream() { delete rdbuf(nullptr); }
+};
+
+NAMESPACE_END(detail)
+
+
+/** \rst
+    Scoped ostream output redirect
+    This a move-only guard that redirects output.
+
+    .. code-block:: cpp
+
+        #include <pybind11/iostream.h>
+
+        int main() {
+            py::scoped_ostream_redirect output;
+            std::cout << "Hello, World!"; // Python stdout
+        } // <-- return std::cout to normal
+
+    You can explicitly pass the c++ stream and the python object, for example to gaurd stderr instead.
+
+    .. code-block:: cpp
+        int main() {
+            py::scoped_ostream_redirect output{std::cerr, py::module::import("sys").attr("stderr")};
+            std::cerr << "Hello, World!";
+        }
+ \endrst */
+
+class scoped_ostream_redirect {
+    std::streambuf * old {nullptr};
+    std::ostream& costream;
+    detail::opythonstream buffer;
+
+public:
+    scoped_ostream_redirect(
+            std::ostream& costream = std::cout,
+            object pyostream = module::import("sys").attr("stdout") )
+        : costream(costream), buffer(pyostream) {
+        old = costream.rdbuf(buffer.rdbuf());
```

Comment:
Seem like the `opythonstream` class isn't even needed. The buffer could just be:
```c++
detail::pythonbuf buffer;
```
And then:
```c++
old = costream.rdbuf(&buffer);
```